parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,480 @@
+            hidden = self.repackage_hidden(hidden)
+            word_weights = output.squeeze().data.exp()
+            if bsz > 1:
+                value, word_idx = torch.max(word_weights, 1)
+            else:
+                value, word_idx = torch.max(word_weights, 0)
+            # mark end indices for items in batch
+            for k in range(word_idx.size(0)):
+                if not done[k]:
+                    if int(word_idx[k]) == self.END_IDX:
+                        done[k] = True
+                        total_done += 1
+            token_list.append(word_idx.view(bsz, 1))
+            i += 1
+
+        return torch.cat(token_list,1)
+
+    def predict(self, data, hidden, targets=None, is_training=True, end_idxs=None, y_lens=None):
+        """Produce a prediction from our model.
+        """
+        loss_dict = None
+        output = None
+        predictions = None
+        if is_training:
+            self.model.train()
+            self.zero_grad()
+            output, hidden = self.model(data, hidden)
+            loss = self.criterion(output.view(-1, len(self.dict)), targets)
+            loss.backward(retain_graph=True)
+            self.update_params()
+            loss_dict = {'lmloss': loss.data}
+            loss_dict['lmppl'] = math.exp(loss.data)
+        else:
+            self.model.eval()
+            predictions = self.get_predictions(data, end_idxs)
+            loss_dict = {}
+            bsz = data.size(0)
+            if bsz != self.batchsize:
+                self.hidden = self.model.init_hidden(bsz)
+            loss = self.get_target_loss(data, self.hidden, targets, y_lens)
+            loss_dict['loss'] = loss
+            loss_dict['ppl'] = math.exp(loss)
+
+        return output, hidden, loss_dict, predictions
+
+    def batchify(self, observations, seq_len, is_training):
+        """Convert a list of observations into input & target tensors."""
+        labels = None
+        valid_inds = None
+        end_idxs = None
+        y_lens = None
+        if is_training:
+            for obs in observations:
+                if obs:
+                    if 'text2vec' in obs:
+                        self.next_batch += obs['text2vec']
+            if len(self.next_batch) <= (seq_len + 1) * self.batchsize:
+                return None, None, None, None, None, None
+            else:
+                data_list = []
+                targets_list = []
+                total = len(self.next_batch)//((seq_len + 1) * self.batchsize)
+                for _ in range(total):
+                    batch = self.next_batch[:((seq_len + 1) * self.batchsize)]
+                    self.next_batch = self.next_batch[((seq_len + 1) * self.batchsize):]
+
+                    source = torch.LongTensor(batch).view(self.batchsize, -1).t().contiguous()
+                    data = Variable(source[:seq_len])
+                    targets = Variable(source[1:].view(-1))
```

Comment:
better to do view before passing into the criterion instead of here--easier to visualize when debugging